File: apps/mobile/lib/utils/ai_base_url.dart
/// Normalizes an OpenAI-compatible API base URL for [AiClient].
///
/// Accepts common user inputs such as the host root, a `/v1` base, or a full
/// `/v1/chat/completions` endpoint pasted by mistake.
String normalizeAiBaseUrl(String raw) {
  var url = raw.trim();
  if (url.isEmpty) return url;

  while (url.endsWith('/')) {
    url = url.substring(0, url.length - 1);
  }

  const chatSuffix = '/chat/completions';
  if (url.endsWith(chatSuffix)) {
    url = url.substring(0, url.length - chatSuffix.length);
    while (url.endsWith('/')) {
      url = url.substring(0, url.length - 1);
    }
  }

  if (!url.endsWith('/v1')) {
    url = '$url/v1';
  }

  return url;
}

File: apps/ios/FreshPantry/Services/AiBaseURL.swift
import Foundation

/// Normalizes an OpenAI-compatible API base URL for `AiClient`.
///
/// Accepts common user inputs such as the host root, a `/v1` base, or a full
/// `/v1/chat/completions` endpoint pasted by mistake (services invariant #11).
/// Ported VERBATIM from `lib/utils/ai_base_url.dart`.
func normalizeAiBaseUrl(_ raw: String) -> String {
    var url = raw.trimmingCharacters(in: .whitespacesAndNewlines)
    if url.isEmpty { return url }

    while url.hasSuffix("/") {
        url.removeLast()
    }

    let chatSuffix = "/chat/completions"
    if url.hasSuffix(chatSuffix) {
        url.removeLast(chatSuffix.count)
        while url.hasSuffix("/") {
            url.removeLast()
        }
    }

    if !url.hasSuffix("/v1") && !url.contains("/v1/") {
        url += "/v1"
    }

    return url
}
